### ISSUE TITLE
ServiceBusTrigger triggers multiple times the same message

### ISSUE THREAD

**Issue by lukasvosyka** [Azure/azure-functions-host#2796]
Hi folks,
I'm now sitting several days on a simple task, that I don't seem to be able to accomplish with Azure Functions (Consumption plan).
I have a 800MB file uploaded to storage and want it to be processed (once!) by a function. Processing time can take up to 7 minutes (using XML Reader, etc).
The problem seems to be, that when the function ends successfully (no exception, no return value), the function starts over with the same input! 
I tried several documented approaches, but am facing sometimes unpredicatable behavior from Azure functions. Also tried different settings in host.json, etc. I observed it on QueueTrigger as well as with ServiceBusTrigger.

When using ServiceBus I see with the help of ServiceBus Explorer (https://github.com/paolosalvatori/ServiceBusExplorer) that after successful processing of the message, the message remains in the queue. After several successful retries (5 per default) it gets into the Dead Letter queue (because I configured the SB to do so). At the end you will find the example Function. Funny fact: When I deploy the Function and trigger a message, AI tells me that it is spinning up server after server (21 by total I got now)...why? I have tried with an empty host.json (so defaults).

The same I observed with QueueTrigger. When using Microsoft Storage Explorer to trigger a message, the message disappears first when fetched by the Function, but after the Function ends, it remains in the Queue (but still invisible... the app's status bar shows "Displaying 0 of 1 messages"). When the Function App is stopped, and then restarted, you can see the message again in the Storage Explorer... after a while it disappears again (due to the lease lock I guess).

So what should I do, if I want to simply have a Function running once per file (it should run in parallel for different files, but once for the very same file)?

#### Investigative information

Please provide the following:

I have attached an extract from AI Analytics that shows the multiple invocation if that helps...

[query_data.zip](https://github.com/Azure/azure-functions-host/files/1975665/query_data.zip)

#### Repro steps

Provide the steps required to reproduce the problem:
Function App on Consumption Plan!

The function for the Service Bus Trigger problem is attached as source code. It is dead simple.
ServiceBusTrigger with a model, Input binding on a Blob, simulate processing (delay 7 minutes). Application insights is configured with the Function and it shows several times, the message about the trigger of the same message. The Function source is at the bottom. Easy to reproduce.

#### Expected behavior

After the Function ends, the message should not be processed again.

#### Actual behavior

After the Function ends, the message gets executed again until the MaxRetry is reached (5 I think is default).

#### Known workarounds

I don't know?

#### Related information 

Provide any related information 

* Programming language used : C#  (Microsoft.NET.Sdk)
* NuGet packages: Microsoft.Azure.WebJobs.ServiceBus (2.2.0)
* Microsof.NET.Sdk.Functions (1.0.13)
* Links to source: I can provide (skype name: morpheus.crc)
* Bindings used: ServiceBusTrigger/QueueTrigger, Blob (input binding)

<details>
<summary>Source</summary>

```csharp
public static class Function1
    {
		[FunctionName("TestLongRunningProcess")]
		[Singleton("{Path}", SingletonScope.Function, Mode = SingletonMode.Listener)]
		public static async Task ProcessEbusUpdatesAndPublishOnEventHub(
			[ServiceBusTrigger("file-change", Microsoft.ServiceBus.Messaging.AccessRights.Listen, Connection = "ServiceBusConnection")] ClassContainingPathProperty trigger,
			[Blob("files/{Path}", FileAccess.Read)] Stream stream,
			ILogger log)
		{
			log.LogInformation("TestLongRunningProcess: {sequenceId} Started with data {storeId}, {fileName}, Stream size {streamSize}", trigger.Sequence, trigger.StoreId, trigger.Path, stream.Length);

			await Task.Delay(TimeSpan.FromMinutes(7));
			log.LogInformation("TestLongRunningProcess: {sequenceId} Ended.", trigger.Sequence);
		}
	}
```
The host.json file
```json
{
   "functionTimeout": "00:10:00"
}
```
</details>

**Comment by mathewc** [comment]
I just tried and this is working fine for me - I can't repro. I tried using a queue (but you said you're seeing this for both SB and Queues). Here's my code:

```csharp
[FunctionName("ProcessQueueMessage")]
[Singleton("{Path}", SingletonScope.Function, Mode = SingletonMode.Listener)]
public static async Task Run(
    [QueueTrigger("repro-queue-item")] Item item, string id,
    [Blob("repro-blobs/{BlobName}", FileAccess.Read)] Stream stream,
    ILogger log)
{
    log.LogInformation($"Processing message: (Id={id}, BlobName={item.BlobName})");

    for (int i = 0; i < 7; i++)
    {
        log.LogInformation("Waiting for a minute...");
        await Task.Delay(TimeSpan.FromMinutes(1));
    }

    using (var sr = new StreamReader(stream))
    {
        var contents = sr.ReadToEnd();
        log.LogInformation(contents);
    }

    log.LogInformation($"Processing complete (Id={id}).");
}
```

And here is my portal output, showing the message was processed, and no additional messages were processed after it:

```
 Pause  Clear   Copy logs   Expand 2018-05-10T00:35:21.964 [Info] Assembly reference changes detected. Restarting host...
2018-05-10T00:35:47.354 [Info] Function started (Id=5f1eadd7-251e-4697-ba3a-e33ef5221b07)
2018-05-10T00:35:47.432 [Info] Processing message: (Id=492abcb1-181b-43dd-9eb8-61e625289db3, BlobName=test.txt
2018-05-10T00:35:47.432 [Info] Waiting for a minute...
2018-05-10T00:36:47.436 [Info] Waiting for a minute...
2018-05-10T00:37:47.441 [Info] Waiting for a minute...
2018-05-10T00:38:47.446 [Info] Waiting for a minute...
2018-05-10T00:39:47.446 [Info] Waiting for a minute...
2018-05-10T00:40:47.462 [Info] Waiting for a minute...
2018-05-10T00:41:47.469 [Info] Waiting for a minute...
2018-05-10T00:42:47.503 [Info] This is a test
2018-05-10T00:42:47.503 [Info] Processing complete (Id=492abcb1-181b-43dd-9eb8-61e625289db3).
2018-05-10T00:42:47.503 [Info] Function completed (Success, Id=5f1eadd7-251e-4697-ba3a-e33ef5221b07, Duration=420142ms)
2018-05-10T00:44:40  No new trace in the past 1 min(s).
```
Are you sure it's the same messages being processed? Can you log their IDs like I'm doing in my example to verify? Another thing that happens to me sometimes that can be confusing is having old invisible failed messages in the queue while I'm actively coding a function. They'll become visible again over time and will be reprocessed again. If those test messages have the same content (e.g. same BlobPath value) I get what looks like duplicate processing, but they're actually different queue messages.

**Comment by lukasvosyka** [comment]
Hi Mathew!
First of all thanks for taking you time on this one. I attached all my Streaming Logs (as there should be the request id's you are asking for) plus my used solution at the end.

So, for me it kinda works for now with the Queue Trigger. As you mention, it gets requeued and processed again, when you are deploying again (I am using V1 Functions and VS to deploy). Maybe that confused me in the first place. But still after testing I am wondering ;-) Maybe you can give me a hint why things are like I observed in the following tests:

**Test 1 - Queue Trigger**
During my tests today, based on your response, I retried the Queue trigger. But I figured a weird behavior and I thought I let you know:
Maybe I first explain, why I am trying to use the Singleton attribute (I want to achive something that seems to be an issue already for a while: https://github.com/Azure/azure-functions-host/issues/912): Let's say I have a model that I am putting into the Queue looking like this:
```javascript
{
	"Sequence": "1",
	"LockContext": "9762",
	"Path": "9762/GK12KASSE.9762.SH.2411201604353686.XML"
}
```
What I am trying to achive is, that no more than one instance of a function is processing files within that subfolder "9762", but if there is a new file in another folder it can be processed at the same time. But within the same folder it should get processed 1 after the other..

The Lock context (from the model above) is the one I am using for the Singleton attribute: 

```csharp
[FunctionName("TestLongRunningProcessQueue")]
[Singleton("{LockContext}", Mode = SingletonMode.Listener)]
public static async Task TestLongRunningProcessQueue(
	[QueueTrigger("ebus-file-change")] Item trigger,
	string id,
	[Blob("ebus-files/{Path}", FileAccess.Read)] Stream stream,
	ILogger log)
{
	log.LogInformation("TestLongRunningProcessQueue (Id={id}: Sequence {sequenceId} started, {fileName}, Stream size {streamSize}", id, trigger.Sequence, trigger.Path, stream.Length);
	for (int i = 0; i < 7; i++)
	{
		log.LogInformation("Waiting for a minute...");
		await Task.Delay(TimeSpan.FromMinutes(1));
	}

	log.LogInformation("TestLongRunningProcessQueue: Sequence {sequenceId} ended.", trigger.Sequence);
}
```

I triggered via Azure Storage Explorer 2 messages. In between I redeployed (the logs attached are from Streaming logs once I deployed). Then I triggered again 2 messages. 
My observation was, that the first two messages (after deployment) have been started (sequence 3 and 4). After a while the previous 2 messages have started being processed (sequence 0 and way later sequence 2). But sequence 2 never finished. Nor can I now add any new message into the queue so they get processed (after finishing my writing here, my queue was empty.. so now I'm not sure what happened). Here are the Logs:
[test1.txt](https://github.com/Azure/azure-functions-host/files/1996183/test1.txt)
The second test (without a deployment in between looked better :))
[test2.txt](https://github.com/Azure/azure-functions-host/files/1996253/test2.txt)

**Test 2 - Service Bus Trigger**
This one is the weird part I don't get why it constantly retriggers (maybe the lock duration configured on the Service Bus?).
So what I did is, that I essentially exchanged the QueueTrigger with ServiceBusTrigger and reran above function and queued 1 single message. I had to purge the queue manually so it does not get reprocessed (after almost 1h of waiting to finish ^^) It behaved completely different:
- It started 41 (!!) servers (as per application insight)
- The Streaming logs show a weird output (it started to process the message, but after about the second log entry "Waiting for a minute..." it stopped logging anything. And there is a "Function started (Id=1c81ce6b-2148-4358-a288-7b31795ea157)" but no "Function complete..." Here is the log from the Streaming logs: 
[test3.txt](https://github.com/Azure/azure-functions-host/files/1996291/test3.txt)
- Attached a screenshot from AI: 
![ai_test_sb](https://user-images.githubusercontent.com/7419311/39939473-b440de3c-5556-11e8-9d16-ae43322e4e3e.PNG)

- And the logs from AI (it's CSV) for this particular test case: 
[ai_test_sb.txt](https://github.com/Azure/azure-functions-host/files/1996374/ai_test_sb.txt)

I hope I can help on this. Attached you will find the solution I was working with:
[LongRunningTest.zip](https://github.com/Azure/azure-functions-host/files/1996381/LongRunningTest.zip)

In case I did something wrong, please let me know. I am a bit confused especially with the SB behavior!

Thank you!
Lukas

**Comment by lukasvosyka** [comment]
Hi Mathew,

> ...you're trying to bind to invocation specific data via {LockContext}, which won't work.

The link you provided tells about "Scoped Values". Are you saying this is this currently not supported for Functions then?

> You mentioned "Sequence 2 never finished". Are you only basing that on the streaming logs not showing the completion?

Yes I was telling that because of that :-)

> However I do see some timeout exceptions because some of your function invocations are exceeding the default execution limit of 5 minutes. See functionTimeout in the wiki here. You can increase this to a max of 10 minutes in the Dynamic plan, no limit for App Service plans. Those timeouts will cause a function failure, and message reprocessing (by design).

All my tests were done with `functionTimeout` set to 00:10:00, as you can see in the attached solution I provided. The logs I believe also contain an entry about that.

> Regarding the high instance count - do you by any chance have any Deferred messages in that queue?

No, I have ensured upfront that the service bus was empty when I started. As mentioned, I only sent a single message (the very same as I did with QueueTrigger). To make life easier with ServiceBus I was using a tool that you can find here: https://github.com/paolosalvatori/ServiceBusExplorer

So, I think at the end QueueTrigger seems to do its Job as designed (I was just confused about the missing entries in the Streaming logs, but AI showed, as you mentioned, the completion of all of them).
But the ServiceBus problem seems to persist and maybe related to the blob binding? I don't see another reason for having OutOfMemory exceptions..

**Comment by mathewc** [comment]
Yes, Singleton scoped values are supported for Functions. In the wiki example note that the scope was set to "Host" not "Listener" (`[Singleton("ItemsLock", SingletonScope.Host)]`).

I'm not sure why the portal logs disconnected and didn't show the completion logs. As long as you keep the portal page open on that invocation and the function is running and producing output, the logs should make it there. If you're not seeing that and have a consistent repro, you can log an issue here: https://github.com/Azure/azure-functions-ux.

For your ServiceBus issue - please do another run that causes your problem, and share the timestamp range so we can investigate. I'm only seeing a few invocations from 5/11 so don't have much to go on. Let's focus on the SB issue at this point - Queues are fine. Also, before the run, fix the Singleton issue.

**Comment by lukasvosyka** [comment]
Now the processing finished and the SB message found its way to the dead letter queue.

Those are the details that might be of interest to you! Please let me know if I can share anymore details that helps you figure the issue.

 | timestamp               | message                                                                                   | opertation id                        |
 | ----------------------- | ----------------------------------------------------------------------------------------- | ------------------------------------ |
 | 2018-05-14T17:52:51.000 | Function started (Id=c5339ee3-cb94-4ed9-b808-cc4cc7e13799)                                | c5339ee3-cb94-4ed9-b808-cc4cc7e13799 |
 | 2018-05-14T17:58:21.231 | Function started (Id=644b3ddc-9bba-461e-813e-6c28ebe59e24)                                | 644b3ddc-9bba-461e-813e-6c28ebe59e24 |
 | 2018-05-14T17:59:51.790 | Function completed (Success, Id=c5339ee3-cb94-4ed9-b808-cc4cc7e13799, Duration=420787ms)  | c5339ee3-cb94-4ed9-b808-cc4cc7e13799 |
 | 2018-05-14T18:03:33.679 | Function started (Id=c58246c7-2f40-48ec-bb69-38b2ebce982f)                                | c58246c7-2f40-48ec-bb69-38b2ebce982f |
 | 2018-05-14T18:06:52.121 | Function completed (Success, Id=644b3ddc-9bba-461e-813e-6c28ebe59e24, Duration=510896ms)  | 644b3ddc-9bba-461e-813e-6c28ebe59e24 |
 | 2018-05-14T18:08:44.163 | Function started (Id=e213e9c1-fafa-455b-8bc4-a47c31768b49)                                | e213e9c1-fafa-455b-8bc4-a47c31768b49 |
 | 2018-05-14T18:13:54.445 | Function started (Id=afd078b6-db21-4113-9e80-d49bbe21cc20)                                | afd078b6-db21-4113-9e80-d49bbe21cc20 |
 | 2018-05-14T18:13:55.233 | Function completed (Success, Id=c58246c7-2f40-48ec-bb69-38b2ebce982f, Duration=621500ms)  | c58246c7-2f40-48ec-bb69-38b2ebce982f |
 | 2018-05-14T18:19:04.788 | Function started (Id=4e5e37e3-416a-47d3-a25a-6fa4967a96ed)                                | 4e5e37e3-416a-47d3-a25a-6fa4967a96ed |
 | 2018-05-14T18:20:59.544 | Function completed (Success, Id=afd078b6-db21-4113-9e80-d49bbe21cc20, Duration=425094ms)  | afd078b6-db21-4113-9e80-d49bbe21cc20 |
 | 2018-05-14T18:24:17.107 | Function started (Id=824af556-1433-4241-9d56-26f35e2f7e6c)                                | 824af556-1433-4241-9d56-26f35e2f7e6c |
 | 2018-05-14T18:28:00.638 | Function completed (Success, Id=4e5e37e3-416a-47d3-a25a-6fa4967a96ed, Duration=535812ms)  | 4e5e37e3-416a-47d3-a25a-6fa4967a96ed |
 | 2018-05-14T18:29:27.337 | Function started (Id=0b4a9fcf-c097-4802-bf87-02075998dc84)                                | 0b4a9fcf-c097-4802-bf87-02075998dc84 |
 | 2018-05-14T18:34:57.600 | Function started (Id=bd0a8532-7033-4bb6-a22e-033a21f87dc8)                                | bd0a8532-7033-4bb6-a22e-033a21f87dc8 |
 | 2018-05-14T18:35:01.938 | Function completed (Success, Id=e213e9c1-fafa-455b-8bc4-a47c31768b49, Duration=1577776ms) | e213e9c1-fafa-455b-8bc4-a47c31768b49 |
 | 2018-05-14T18:40:10.973 | Function started (Id=ac5d2856-4f56-4c7a-8140-6a49cd83695f)                                | ac5d2856-4f56-4c7a-8140-6a49cd83695f |
 | 2018-05-14T18:42:03.198 | Function completed (Success, Id=bd0a8532-7033-4bb6-a22e-033a21f87dc8, Duration=425585ms)  | bd0a8532-7033-4bb6-a22e-033a21f87dc8 |
 | 2018-05-14T18:49:04.171 | Function completed (Success, Id=0b4a9fcf-c097-4802-bf87-02075998dc84, Duration=1176827ms) | 0b4a9fcf-c097-4802-bf87-02075998dc84 |
 | 2018-05-14T18:56:05.849 | Function completed (Success, Id=824af556-1433-4241-9d56-26f35e2f7e6c, Duration=1908753ms) | 824af556-1433-4241-9d56-26f35e2f7e6c |
 | 2018-05-14T19:03:08.660 | Function completed (Success, Id=ac5d2856-4f56-4c7a-8140-6a49cd83695f, Duration=1377725ms) | ac5d2856-4f56-4c7a-8140-6a49cd83695f |

**Comment by lukasvosyka** [comment]
I have now tried without the Singleton attribute at all. At the first glance it behaves the very same way...BUT I think I see a pattern:

Each of those reinvoked messages happen exactly after 00:05:30.. Seems like the service bus trigger waits for a constant time until the execution host tries to reprocess the message, despite the fact, that the functionTimeout has been set to 10mins? I remember that Functions Invocations were hard-coded to 5 mins max. execution time and this limit was softened. Maybe this hasn't reached the SB trigger implementation? ;-D

Here are details of my second test case running:
 | timestamp               | message                                                                                   | opertation id                        |
 | ----------------------- | ----------------------------------------------------------------------------------------- | ------------------------------------ |
 | 2018-05-14T19:57:11.237 | Function started (Id=25c5da71-d2e2-42ce-bcbe-b0aa585a689f)                                | 25c5da71-d2e2-42ce-bcbe-b0aa585a689f |
 | 2018-05-14T20:02:41.400 | Function started (Id=cbc5f8b8-0b5c-453e-8fec-7edf983cedf8)                                | cbc5f8b8-0b5c-453e-8fec-7edf983cedf8 |
 | 2018-05-14T20:04:11.886 | Function completed (Success, Id=25c5da71-d2e2-42ce-bcbe-b0aa585a689f, Duration=420639ms)  | 25c5da71-d2e2-42ce-bcbe-b0aa585a689f |
 | 2018-05-14T20:07:52.694 | Function started (Id=2d5227a6-3b23-48a6-98ab-5427514b9ca4)                                | 2d5227a6-3b23-48a6-98ab-5427514b9ca4 |
 | 2018-05-14T20:09:42.004 | Function completed (Success, Id=cbc5f8b8-0b5c-453e-8fec-7edf983cedf8, Duration=420599ms)  | cbc5f8b8-0b5c-453e-8fec-7edf983cedf8 |

**Comment by lukasvosyka** [comment]
After some further investigation it seems to me like the auto renewal of the lock doesn't work properly? The default is set to 5 mins, but a 7 mins running function does not seem to renew the lock, hence the message gets reprocessed. Can you confirm?

**Comment by lukasvosyka** [comment]
Ok thanks Mathew for confirming my guess :-D
Still, it seems being a bug then, no? When a function is allowed to run 7 mins, the host.json has the config option to set an autoRenewTimeout and still not working as intended.

For myself, I found a way I guess to move on! Thanks also for the hint on the correct use of the Singleton attribute. As soon as I stumble across the next "weird" behavior I'll post again ;-)

Thanks again!
Lukas

**Comment by lukasvosyka** [comment]
Jep, that did the trick! Thanks for the link to sfeldman's blog. It helped me understand what the autoRenewTimeout stands for... I was kinda expecting that a lock renewal happens after the timeout, until the message gets processed... but it actually stands for a lock duration. The docs on the https://github.com/Azure/azure-functions-host/wiki/host.json page don't tell that so explicit. Now I see, thanks!

**Comment by rimbakasimba** [comment]
@lukasvosyka  you missed pointing to the link to the blog you referred.. https://weblogs.asp.net/sfeldman/azure-service-bus-autorenewtimeout
